sklearn/ensemble/iforest.py

```diff
@@ -246,14 +266,51 @@ def decision_function(self, X):
 
         Returns
         -------
-        scores : array of shape (n_samples,)
+        scores : array, shape (n_samples,)
             The anomaly score of the input samples.
-            The lower, the more abnormal.
+            The lower, the more abnormal. Negative scores represent outliers,
+            positive scores represent inliers.
+
+        """
+        # We substract self.offset_ to make 0 be the threshold value for being
+        # an outlier:
+
+        return self.score_samples(X) - self.offset_
+
+    def score_samples(self, X):
+        """Opposite of the anomaly score defined in the original paper.
+
+        The anomaly score of an input sample is computed as
+        the mean anomaly score of the trees in the forest.
+
+        The measure of normality of an observation given a tree is the depth
+        of the leaf containing this observation, which is equivalent to
+        the number of splittings required to isolate this point. In case of
+        several observations n_left in the leaf, the average path length of
+        a n_left samples isolation tree is added.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            The training input samples. Sparse matrices are accepted only if
+            they are supported by the base estimator.
 
+        Returns
+        -------
+        scores : array, shape (n_samples,)
+            The anomaly score of the input samples.
+            The lower, the more abnormal.
         """
         # code structure from ForestClassifier/predict_proba
+        check_is_fitted(self, ["estimators_"])
+
         # Check data
         X = check_array(X, accept_sparse='csr')
+        if self.n_features_ != X.shape[1]:
+            raise ValueError("Number of features of the model must "
```

Comment:
Not currently tested. A common test for outliers, or for `score_samples` (perhaps just extending similar checks in `estimator_checks.py`)

Reply:
I simply added a test to test_iforest but I agree we should write common tests for outliers detection algo when the API will be shared (ie after making `predict` public for LOF)

Reply:
There is already a PR (#9270) implementing common tests for outlier detection estimators.

I ran the common tests of PR #9270 for this PR a few months ago and I think this is the reason why @ngoix added a 'raise ValueError' for such a case because one of the common tests was failing (see [my previous review](https://github.com/scikit-learn/scikit-learn/pull/9015#pullrequestreview-64941309)). Hence such a common test should already be implemented in PR #9270. I will finish PR #9270 once this PR is merged.